mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala

```diff
@@ -206,6 +209,29 @@ class SymbolIndexBucket(
         .getOrElse(List.empty)
     }
   }
+
+  /**
+   * Remove possible old, outdated entries from the toplevels and definitions.
+   * This action is performed when a symbol is queried, to avoid returning incorrect results.
+   */
+  private def removeOldEntries(symbol: Symbol): Unit = {
+    toplevels.get(symbol.value) match {
+      case None => ()
+      case Some(acc) =>
+        val updated = acc.filter(_.exists)
+        if (updated.isEmpty) toplevels.remove(symbol.value)
+        else toplevels(symbol.value) = updated
+    }
+
+    definitions.get(symbol.value) match {
+      case None => ()
+      case Some(acc) =>
+        val updated = acc.filter(_.path.exists)
+        if (updated.isEmpty) definitions.remove(symbol.value)
+        else definitions(symbol.value) = updated
```

Comment:
Maybe, let's not update it if no files were removed from acc by the filter? The same above.

Reply:
we can do that, absolutely. I was also thinking about storing and passing set with (removed, existing) paths so we can reduce number of IO, wdyt?

Reply:
I actually had a bug that this would fix :sweat_smile: I'll merge it and update

Reply:
What bug did you have?

Reply:
I changed branches and go to definition was using a stale location

Reply:
```
ERROR searching for `scala/meta/internal/semver/SemVer.Version.fromString().` failed
java.nio.file.NoSuchFileException: /home/tgodzik/Documents/metals/mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala
at [sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fusr%2Flib%2Fjvm%2Fopenjdk17%2Flib%2Fsrc.zip%21%2Fjava.base%2Fsun%2Fnio%2Ffs%2FUnixException.java%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A91%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A91%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fusr%2Flib%2Fjvm%2Fopenjdk17%2Flib%2Fsrc.zip%21%2Fjava.base%2Fsun%2Fnio%2Ffs%2FUnixException.java%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A105%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A105%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fusr%2Flib%2Fjvm%2Fopenjdk17%2Flib%2Fsrc.zip%21%2Fjava.base%2Fsun%2Fnio%2Ffs%2FUnixException.java%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A110%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A110%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:218)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fusr%2Flib%2Fjvm%2Fopenjdk17%2Flib%2Fsrc.zip%21%2Fjava.base%2Fsun%2Fnio%2Ffs%2FUnixFileSystemProvider.java%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A217%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A217%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [java.nio.file.Files.newByteChannel(Files.java:380)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fusr%2Flib%2Fjvm%2Fopenjdk17%2Flib%2Fsrc.zip%21%2Fjava.base%2Fjava%2Fnio%2Ffile%2FFiles.java%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A379%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A379%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [java.nio.file.Files.newByteChannel(Files.java:432)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fusr%2Flib%2Fjvm%2Fopenjdk17%2Flib%2Fsrc.zip%21%2Fjava.base%2Fjava%2Fnio%2Ffile%2FFiles.java%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A431%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A431%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [java.nio.file.Files.readAllBytes(Files.java:3288)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fusr%2Flib%2Fjvm%2Fopenjdk17%2Flib%2Fsrc.zip%21%2Fjava.base%2Fjava%2Fnio%2Ffile%2FFiles.java%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A3287%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A3287%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.io.PlatformFileIO$.slurp(PlatformFileIO.scala:45)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fhome%2Ftgodzik%2F.cache%2Fcoursier%2Fv1%2Fhttps%2Frepo1.maven.org%2Fmaven2%2Forg%2Fscalameta%2Ftrees_2.13%2F4.7.1%2Ftrees_2.13-4.7.1-sources.jar%21%2Fscala%2Fmeta%2Finternal%2Fio%2FPlatformFileIO.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A44%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A44%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.io.FileIO$.slurp(FileIO.scala:24)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fhome%2Ftgodzik%2F.cache%2Fcoursier%2Fv1%2Fhttps%2Frepo1.maven.org%2Fmaven2%2Forg%2Fscalameta%2Ftrees_2.13%2F4.7.1%2Ftrees_2.13-4.7.1-sources.jar%21%2Fscala%2Fmeta%2Finternal%2Fio%2FFileIO.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A23%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A23%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.ScalametaCommonEnrichments$XtensionAbsolutePath.toInput(ScalametaCommonEnrichments.scala:380)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FScalametaCommonEnrichments.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A379%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A379%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.SymbolIndexBucket.addMtagsSourceFile(SymbolIndexBucket.scala:216)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FSymbolIndexBucket.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A215%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A215%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.SymbolIndexBucket.$anonfun$query0$1(SymbolIndexBucket.scala:181)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FSymbolIndexBucket.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A180%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A180%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.SymbolIndexBucket.$anonfun$query0$1$adapted(SymbolIndexBucket.scala:181)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FSymbolIndexBucket.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A180%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A180%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.collection.immutable.Set$Set1.foreach(Set.scala:168)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fhome%2Ftgodzik%2F.cache%2Fcoursier%2Fv1%2Fhttps%2Frepo1.maven.org%2Fmaven2%2Forg%2Fscala-lang%2Fscala-library%2F2.13.10%2Fscala-library-2.13.10-sources.jar%21%2Fscala%2Fcollection%2Fimmutable%2FSet.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A167%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A167%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.SymbolIndexBucket.query0(SymbolIndexBucket.scala:181)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FSymbolIndexBucket.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A180%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A180%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.SymbolIndexBucket.query(SymbolIndexBucket.scala:158)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FSymbolIndexBucket.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A157%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A157%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.OnDemandSymbolIndex.$anonfun$findSymbolDefinition$1(OnDemandSymbolIndex.scala:139)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FOnDemandSymbolIndex.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A138%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A138%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.collection.immutable.List.flatMap(List.scala:293)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22jar%3Afile%3A%2F%2F%2Fhome%2Ftgodzik%2F.cache%2Fcoursier%2Fv1%2Fhttps%2Frepo1.maven.org%2Fmaven2%2Forg%2Fscala-lang%2Fscala-library%2F2.13.10%2Fscala-library-2.13.10-sources.jar%21%2Fscala%2Fcollection%2Fimmutable%2FList.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A292%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A292%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.OnDemandSymbolIndex.findSymbolDefinition(OnDemandSymbolIndex.scala:139)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FOnDemandSymbolIndex.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A138%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A138%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at [scala.meta.internal.mtags.OnDemandSymbolIndex.definitions(OnDemandSymbolIndex.scala:56)](command:metals.metals-goto-location?%5B%7B%22uri%22%3A%22file%3A%2F%2F%2Fhome%2Ftgodzik%2FDocuments%2Fmetals%2Fmtags%2Fsrc%2Fmain%2Fscala%2Fscala%2Fmeta%2Finternal%2Fmtags%2FOnDemandSymbolIndex.scala%22%2C%22range%22%3A%7B%22start%22%3A%7B%22line%22%3A55%2C%22character%22%3A0%7D%2C%22end%22%3A%7B%22line%22%3A55%2C%22character%22%3A0%7D%7D%2C%22otherWindow%22%3Atrue%7D%5D)
at scala.meta.internal.metals.DestinationProvider.de
```